simple_metering.cpp

```diff
@@ -0,0 +1,213 @@
+            }
+                break;
+
+            default:
+                break;
+            }
+        }
+
+        if (stateUpdated)
+        {
+            sensor->updateStateTimestamp();
+            enqueueEvent(Event(RSensors, RStateLastUpdated, sensor->id()));
+        }
+
+        if (configUpdated || stateUpdated)
+        {
+            updateSensorEtag(&*sensor);
+            sensor->setNeedSaveDatabase(true);
+            queSaveDb(DB_SENSORS, DB_SHORT_SAVE_DELAY);
+        }
+    }
+}
+
+/*! Write Attribute on thermostat cluster.
+ *  Iterate over every day and get schedule for each day.
+   \param task - the task item
+   \param attrId
+   \param attrType
+   \param attrValue
+   \return true - on success
+           false - on error
+ */
+bool DeRestPluginPrivate::addTaskSimpleMeteringReadWriteAttribute(TaskItem &task, uint8_t readOrWriteCmd, uint16_t attrId, uint8_t attrType, uint32_t attrValue, uint16_t mfrCode)
+{
+    if (readOrWriteCmd != deCONZ::ZclReadAttributesId && readOrWriteCmd != deCONZ::ZclWriteAttributesId)
+    {
+        DBG_Printf(DBG_INFO, "Invalid command for simple metering cluster %d\n", readOrWriteCmd);
+        return false;
+    }
+
+    task.taskType = TaskSimpleMetering;
+
+    task.req.setClusterId(METERING_CLUSTER_ID);
+    task.req.setProfileId(HA_PROFILE_ID);
+
+    task.zclFrame.payload().clear();
+    task.zclFrame.setSequenceNumber(zclSeq++);
+    task.zclFrame.setCommandId(readOrWriteCmd);
+    task.zclFrame.setFrameControl(deCONZ::ZclFCProfileCommand |
+            deCONZ::ZclFCDirectionClientToServer |
+            deCONZ::ZclFCDisableDefaultResponse);
+
+    if (mfrCode != 0x0000)
+    {
+        task.zclFrame.setFrameControl(task.zclFrame.frameControl() | deCONZ::ZclFCManufacturerSpecific);
+        task.zclFrame.setManufacturerCode(mfrCode);
+    }
+
+    // payload
+    QDataStream stream(&task.zclFrame.payload(), QIODevice::WriteOnly);
+    stream.setByteOrder(QDataStream::LittleEndian);
+
+    stream << static_cast<quint16>(attrId);
+
+    if (readOrWriteCmd == deCONZ::ZclWriteAttributesId)
+    {
+        stream << static_cast<quint8>(attrType);
+        if (attrType == deCONZ::Zcl8BitUint || attrType == deCONZ::Zcl8BitEnum || attrType == deCONZ::Zcl8BitBitMap || attrType == deCONZ::ZclBoolean)
+        {
+            stream << static_cast<quint8>(attrValue);
+        }
+        else if (attrType == deCONZ::Zcl8BitInt)
+        {
+            stream << static_cast<qint8>(attrValue);
+        }
+        else if (attrType == deCONZ::Zcl16BitUint || attrType == deCONZ::Zcl16BitBitMap || attrType == deCONZ::Zcl16BitEnum)
+        {
+            stream << static_cast<quint16>(attrValue);
+        }
+        else if (attrType == deCONZ::Zcl16BitInt)
+        {
+            stream << static_cast<qint16>(attrValue);
+        }
+        else
+        {
+            return false;
+        }
+    }
```

Comment:
I think the following should work for the given data types:

```cpp

stream << attrId;

if (readOrWriteCmd == deCONZ::ZclWriteAttributesId)
{
    deCONZ::ZclAttribute attr(attrId, attrType, "", deCONZ::ZclWrite, true);
    attr.setValue(QVariant(attrValue));

    if (!attr.writeToStream(stream))
    {
         return false;
    }
}
```

Reply:
Ah, yes. Way more elegant. However, the code must be slightly amended. Tested and working locally.